packages/hydrogen-react/src/RichText.tsx

```diff
@@ -0,0 +1,212 @@
+import {createElement, type ReactNode} from 'react';
+import type {RichTextASTNode} from './RichText.types.js';
+import {
+  type CustomComponents,
+  RichTextComponents,
+} from './RichText.components.js';
+
+export interface RichTextPropsBase<ComponentGeneric extends React.ElementType> {
+  /** An HTML tag or React Component to be rendered as the base element wrapper. The default is `div`. */
+  as?: ComponentGeneric;
+  /** An object with fields that correspond to the Storefront API's [RichText format](https://shopify.dev/docs/apps/custom-data/metafields/types#rich-text-formatting). */
+  data: RichTextASTNode;
+  /** Customize how rich text components are rendered */
+  components?: CustomComponents;
+  /** Remove rich text formatting and render plain text */
+  plain?: boolean;
+}
+
+export function RichText<ComponentGeneric extends React.ElementType = 'div'>({
+  as,
+  data,
+  plain,
+  components,
+  ...passthroughProps
+}: RichTextProps<ComponentGeneric>): JSX.Element {
+  const Wrapper = as ?? 'div';
+
+  return (
+    <Wrapper {...passthroughProps}>
+      {plain
+        ? richTextToString(data)
+        : serializeRichTextASTNode(components, data)}
+    </Wrapper>
+  );
+}
+
+// This article helps understand the typing here https://www.benmvp.com/blog/polymorphic-react-components-typescript/ Ben is the best :)
+export type RichTextProps<ComponentGeneric extends React.ElementType> =
+  RichTextPropsBase<ComponentGeneric> &
+    Omit<
+      React.ComponentPropsWithoutRef<ComponentGeneric>,
+      keyof RichTextPropsBase<ComponentGeneric>
+    >;
+
+function serializeRichTextASTNode(
+  components: CustomComponents = {},
+  node: RichTextASTNode,
+  index = 0,
+): ReactNode {
+  let children;
+  if ('children' in node) {
+    children = node.children.map((child) =>
+      serializeRichTextASTNode(components, child, index),
+    );
+  }
+
+  const Component =
+    components[node.type === 'list-item' ? 'listItem' : node.type] ??
```

Comment:
Follow the conventions of React to switch to camelCase. This makes it so that customizing the list item looks like:

```ts
import {RichText} from '@shopify/hydrogen';

export function MainRichText({text}: {text: string}) {
  return (
    <RichText
      data={JSON.parse(text)}
      components={{
        listItem({node}) {
          return <li className="customClass">{node.children}</li>;
        },
      }}
    />
  );
}
```